allennlp/semparse/worlds/world.py

```diff
@@ -0,0 +1,519 @@
+from typing import List, Dict, Set, Tuple
+from collections import defaultdict
+import logging
+import re
+
+from nltk.sem.logic import ApplicationExpression, Expression, LambdaExpression, BasicType, Type
+
+from allennlp.semparse.type_declarations import type_declaration as types
+from allennlp.semparse import util as semparse_util
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+class ParsingError(Exception):
+    """
+    This exception gets raised when there is a parsing error during logical form processing.  This
+    might happen because you're not handling the full set of possible logical forms, for instance,
+    and having this error provides a consistent way to catch those errors and log how frequently
+    this occurs.
+    """
+    def __init__(self, message):
+        super(ParsingError, self).__init__()
+        self.message = message
+
+    def __str__(self):
+        return repr(self.message)
+
+
+class ExecutionError(Exception):
+    """
+    This exception gets raised when you're trying to execute a logical form that your executor does
+    not understand. This may be because your logical form contains a function with an invalid name
+    or a set of arguments whose types do not match those that the fuction expects.
+    """
+    def __init__(self, message):
+        super(ExecutionError, self).__init__()
+        self.message = message
+
+    def __str__(self):
+        return repr(self.message)
+
+
+class _TreeNode:
```

Comment:
Can't you use `nltk.Tree` for this? Also it's unintuitive that you don't pass the children into the constructor.

Reply:
I managed to get this to work, though I'm not sure I like it better this way.  `nltk.Tree` inherits from `list`, and so it was un-intuitive to figure out how to build the tree correctly.  We're building the tree incrementally from an action sequence, so we don't know a node's children when we construct the node.  But, `nltk.Tree` has some useful functions, so maybe it's worth it.